tests.sh

```diff
@@ -176,27 +202,42 @@ tests_debug() {
         echo "### ${@}"
     fi >&2
 }
-# }}}
 
-# Internal Code {{{
+# Function tests_do evaluates specified string
+# Args:
+#   ${@}: string to evaluate
 tests_do() {
     tests_debug "$ ${@}"
 
+    echo >$TEST_STDOUT
+    echo >$TEST_STDERR
+    echo >$TEST_EXITCODE
+
     {
         if [ $TEST_VERBOSE -lt 2 ]; then
-            tests_eval "${@}" > /dev/null 2>/dev/null
+            tests_eval "${@}" > $TEST_STDOUT 2> $TEST_STDERR
         fi
 
         if [ $TEST_VERBOSE -lt 3 ]; then
             tests_eval "${@}" > /dev/null
+            tests_eval "${@}" \
+                2> >(tee $TEST_STDERR 1>&2) \
+                1> >(tee $TEST_STDOUT > /dev/null)
         fi
 
         if [ $TEST_VERBOSE -ge 4 ]; then
-            tests_eval "${@}"
+            tests_eval "${@}" \
+                2> >(tee $TEST_STDERR 1>&2) \
```

Comment:
BTW, it's all redirected to stdout later (235 line), so it's useless to redirect it to stderr there.